07.ls_object/formatter.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+require_relative 'file_detail'
+
+class Formatter
+  COL = 3
+
+  def initialize(names)
+    @names = names
+  end
+
+  def format_short
+    row_count = @names.size.fdiv(COL)
+    max_bytesize = @names.max_by(&:bytesize).bytesize
+    cols = @names.each_slice(row_count.ceil)
+    lines = Array.new(row_count.ceil) do |idx|
+      cols.map do |col|
+        col[idx].ljust(max_bytesize)
+      end
+    end
+    lines.each do |name|
+      puts name.join(' ')
+    end
+  end
+
+  def format_long
+    puts "total #{file_details.map(&:blocks).sum}"
+    @names.each do |name|
+      file_detail = file_details.find { |file| file.name == name }
+      nlink = file_detail.nlink.to_s.rjust(max_size(:nlink))
+      owner = file_detail.owner.ljust(max_size(:owner) + 1)
+      group = file_detail.group.ljust(max_size(:group) + 1)
+      size = file_detail.size.to_s.rjust(max_size(:size))
+      month = format('%2d', file_detail.timestamp.month)
+      day = file_detail.timestamp.strftime('%e')
+      time = file_detail.timestamp.strftime('%H:%M')
+      puts [file_detail.mode, nlink, owner, group, size, month, day, time, name].join(' ')
+    end
+  end
+
+  private
+
+  def file_details
+    @file_details = @names.map { |name| FileDetail.new(name) } unless instance_variable_defined?(:@file_details)
+    @file_details
+  end
+
+  def max_size(detail)
+    file_details.map { |file| file.send(detail).to_s.size }.max
```

Comment:
```suggestion
    file_details.map { |file| sleep 0.5; file.send(detail).to_s.size }.max
```

みたいにすると、まためちゃくちゃ遅くなりませんか？(＝今の方法だと処理効率が悪い)

Reply:
めちゃくちゃ遅くなります。
file_detailメソッドは変数に修正して、```FileDetail.new```が一度しか行われないようにしました。これで、```.map```の前に何度も```FileDetail.new```されなくなったので、```file_details.map { |file| sleep 0.5; file.send(detail).to_s.size }.max``` も軽くなるのでは？と考えていたのですが、遅いままです。
Threadによる並列化も試したのですが、遅いままでした。

file_detailsの要素数×0.5秒かかることを回避するためには、```{}``` の中身でなく、```file_details.map``` の部分（）を修正すべきという理解であってますでしょうか？
具体的な方法が思いついてなくて方向性の質問になり申し訳ありません🙇🏻‍♀️

Reply:
これ、先に答えを知っていると「なんだ、それだけか！」という話なのですが、難しく考え出すとどこまでも難しく考えちゃうかもしれませんね。
とりあえず、マルチスレッド化みたいな大変なことはやりません。
ロジックをちょこっと見直すだけです。

たとえば、こんな感じで、毎回ループのたびに`max_size(:nlink)`の結果を出力してみてください。
(sleep はいったん外して大丈夫です）

```ruby
    @file_details.each do |file_detail|
      name = file_detail.name
      puts max_size(:nlink) # ループするたびに結果を出力
      nlink = file_detail.nlink.to_s.rjust(max_size(:nlink))
      # ..
    end
```

結果がわかったら、nlink以外の結果もチェックしてみましょう。
そうすると、改善ポイントが見つかるのでは？

さらにヒント：
sleepを入れた場合でも、totalからその次の行（＝ファイル一覧の1行目）が表示するまでは時間がかかるが、そこからうしろ（＝ファイルの一覧の2行目以降）は一瞬で表示される、というのが理想形です。

Reply:
ヒントありがとうございます！
```.each``` の中でmax_sizeメソッドを呼び出していることで、sleepの呼び出し回数が増えてしまっていることを理解しました🙇🏻‍♀️
```.each``` の外でmax_sizeメソッドを呼び出す形で修正しました。